source/ext/patches/fluentd/in_syslog.patch

```diff
@@ -9,12 +37,79 @@
  
      def configure(conf)
        super
-@@ -251,7 +253,7 @@
+@@ -226,7 +232,7 @@
+         end
+ 
+         ## from receive_data_default
+-        pri = record.delete('pri'.freeze) unless pri
++        pri = record.delete('pri') unless pri
+ 
+         facility = FACILITY_MAP[pri >> 3]
+         priority = PRIORITY_MAP[pri & 0b111]
+@@ -244,14 +250,23 @@
+     private
+ 
+     def listen(callback)
+-      log.info "listening syslog socket on #{@bind}:#{@port} with #{@protocol_type}"
++
+       if @protocol_type == :udp
++        log.info "listening syslog socket on #{@bind}:#{@port} with #{@protocol_type}"
+         @usock = SocketUtil.create_udp_socket(@bind)
+         @usock.bind(@bind, @port)
          SocketUtil::UdpHandler.new(@usock, log, @message_length_limit, callback, @resolve_hostname)
-       else
+-      else
++      elsif @protocol_type == :tcp
++        log.info "listening syslog socket on #{@bind}:#{@port} with #{@protocol_type}"
          # syslog family add "\n" to each message and this seems only way to split messages in tcp stream
 -        Coolio::TCPServer.new(@bind, @port, SocketUtil::TcpHandler, log, "\n", callback, @resolve_hostname)
 +        Coolio::TCPServer.new(@bind, @port, SocketUtil::TcpHandler, log, @delimiter, callback, @resolve_hostname)
++      elsif @protocol_type == :unix
++        if File.exist?(@path)
++          File.unlink(@path)
++        end
++        FileUtils.mkdir_p File.dirname(@path)
++        log.info "listening syslog unix socket #{@path}"
++        Coolio::UNIXServer.new(@path, UnixHandler, log, @delimiter, callback)
        end
      end
  
+@@ -260,5 +275,39 @@
+     rescue => e
+       log.error "syslog failed to emit", error: e.to_s, error_class: e.class.to_s, tag: tag, record: Yajl.dump(record)
+     end
++
++    class UnixHandler < Coolio::Socket
++      def initialize(io, log, delimiter, callback)
++        super(io)
++        @delimiter = delimiter
++        @callback = callback
++        @log = log
++        @buffer = "".force_encoding('ASCII-8BIT')
++      end
++
++      def on_connect
++        @log.trace { "Connected syslog socket object_id=#{self.object_id}" }
++      end
++
++      def on_read(data)
++        @buffer << data
++        pos = 0
++
++        while i = @buffer.index(@delimiter, pos)
++          msg = @buffer[pos...i]
++          @callback.call(msg, @addr)
++          pos = i + @delimiter.length
++        end
++        @buffer.slice!(0, pos) if pos > 0
```

Comment:
whats the point of doing this here, outside the loop

Reply:
That line ensure to not clear buffer entirely, only from 0 to last read position

Reply:
ah ok. so we clear up to the last read position AKA the last delimiter in the input so far

Reply:
yes